docs: add guidance on langsmith deployment scaling (#1393)
langchain-ai/docs: src/langsmith/agent-server-scale.mdx

- If your assistant is CPU bounded, the default value of 10 is likely sufficient. You might lower N_JOBS_PER_WORKER if you notice excessive CPU usage on queue workers or delays in run execution.
- If your assistant is IO bounded, increase N_JOBS_PER_WORKER to handle more concurrent runs per worker.

There is no upper limit to N_JOBS_PER_WORKER. However, Queue workers are greedy when fetching new runs, meaning they will try to pick up as many runs they have available jobs and begin executing them right away. If N_JOBS_PER_WORKER is too high and traffic is burstly, you may see uneven worker utilization and potentially longer run execution times.

#### Avoid synchronous blocking operations

Avoid synchronous blocking operations in your code. If required, set [BG_JOB_ISOLATED_LOOPS](/langsmith/env-var#bg-job-isolated-loops) to True to execute each run in a separate event loop and avoid blocking the main event loop.

#### Minimize redundant checkpointing

Minimize redundant checkpointing by setting [`durability`](/oss/langgraph/durable-execution#durability-modes) to the minimum value necessary to ensure your data is durable.

#### Self-hosted

<Note>

> These settings are only required for self-hosted deployments. By default, Cloud deployments already have these best practices enabled.

Reviewer: ```suggestion
These settings are only required for [self-hosted](/langsmith/self-hosted) deployments. By default, [cloud](/langsmith/cloud) deployments already have these best practices enabled.
```

</Note>

###### Enable the use of queue workers

> By default, the api server manages the queue and does not use queue workers. You can enable the use of queue workers by setting the `queue.enabled` configuration to `true`.

Reviewer: ```suggestion
By default, the API server manages the queue and does not use queue workers. You can enable the use of queue workers by setting the `queue.enabled` configuration to `true`.
```
When we say API server, are we talking about the Agent Server? (This is more a question for me, btw!)

- N_JOBS_PER_WORKER: The number of runs that can be executed at a time by a single queue worker. Defaults to 10.

You can calculate the available jobs with the following equation:
```
available_jobs = number_of_queue_workers * N_JOBS_PER_WORKER
```

Throughput is then the number of runs that can be executed per second by the available jobs:
```
throughput_per_second = available_jobs / average_run_execution_time_seconds
```

So the minimum number of queue workers you should provision to support your expected steady state throughput is:
```
number_of_queue_workers = throughput_per_second * average_run_execution_time_seconds / N_JOBS_PER_WORKER
```

##### Configure autoscaling for bursty workloads

Autoscaling is disabled by default, but should be configured for bursty workloads. Using the same calculations as above, you can determine the maximum number of queue workers you should allow the autoscaler to scale to based on maximum expected throughput.

## Scaling for read load

Read load is primarily driven by the following factors:

- Getting the results of a run
- Getting the state of a thread
- Searching for runs, threads, cron jobs and assistants

> - Retrieving checkpoints and long term memory

Reviewer: Could you add the same set of links here, in case a user lands in this section rather than going from the top of the page to the bottom?

The following components are primarily responsible for handling read load:

> - API server: Handles the request and direct retrieval of data from the database.

Reviewer: I didn't mention this in the previous list of components in the write load section, but do you think it's worthwhile linking to these components in some of the architecture docs we have?

#### Set a TTL on threads to automatically delete old data

Set a TTL on threads to automatically clean up old data. Runs and checkpoints are automatically deleted when the associated thread is deleted.

#### Avoid polling and use /join to monitor the state of a run

Avoid polling the state of a run by using the `/join` API. This API allows you to monitor the state of a run and return a response with the final state once the run is complete.

If you need to monitor the state of a run in real-time, you can use the `/stream` API instead.

#### Self-hosted

<Note>
These settings are only required for self-hosted deployments. By default, Cloud deployments already have these best practices enabled.
</Note>

##### Configure autoscaling for bursty workloads

Autoscaling is disabled by default, but should be configured for bursty workloads. You can determine the maximum number of api servers you should allow the autoscaler to scale to based on maximum expected throughput.

## Example self-hosted Agent Server configurations

<Note>

> The exact optimal configuration depends on your application complexity, request patterns, and data requirements. Use the examples below in combination with the information above and your specific usage to update your deployment configuration as needed. If you have any questions, please reach out to the LangChain team.

Reviewer: ```suggestion
The exact optimal configuration depends on your application complexity, request patterns, and data requirements. Use the following examples in combination with the information in the previous sections and your specific usage to update your deployment configuration as needed. If you have any questions, reach out to the LangChain team at [support@langchain.dev](mailto:support@langchain.dev).
```

| :--- | :--- | :--- | :--- | :--- | :--- |
| <Tooltip tip="Number of write requests being processed by the Deployment per second">Write requests per second</Tooltip> | 5 | 5 | 500 | 50 | 500 |
| <Tooltip tip="Number of read requests being processed by the Deployment per second">Read requests per second</Tooltip> | 5 | 500 | 5 | 50 | 500 |
| **API servers**<br />(1 CPU, 2Gi per server) | 1 (default) | 6 | 10 | 3 | 15 |
| **Queue workers**<br />(1 CPU, 2Gi per worker) | 1 (default) | 10 | 1 (default) | 5 | 10 |
| **N_JOBS_PER_WORKER** | 10 (default) | 50 | 10 | 10 | 50 |
| **Redis resources** | 2 Gi (default) | 2 Gi (default) | 2 Gi (default) | 2 Gi (default) | 2 Gi (default) |
| **Postgres resources** | 2 CPU<br />8 Gi (default) | 4 CPU<br />16 Gi memory | 4 CPU<br />16 Gi | 4 CPU<br />16 Gi memory | 8 CPU<br />32 Gi memory |

Below we provide sample configurations to enable each of these setups. We define each load level as follows:

- Low means roughly 5 requests per second
- Medium means roughly 50 requests per second
- High means roughly 500 requests per second

### Low reads, low writes <a name="low-reads-low-writes"></a>

> The default LangSmith Deployment configuration will handle this load. No custom resource configuration is needed here.

Reviewer: ```suggestion
The default [LangSmith Deployment](/langsmith/deployments) configuration will handle this load. No custom resource configuration is needed here.
```
